crates/subspace-networking/src/protocols/subspace_connection_limits.rs

```diff
@@ -0,0 +1,200 @@
+            incoming_allow_list: HashMap::default(),
+            outgoing_allow_list: HashMap::default(),
+        }
+    }
+
+    /// Add to allow list some attempts of incoming connections from specified peer ID that will bypass global limits
+    pub(crate) fn add_to_incoming_allow_list(&mut self, peer: PeerId, attempts: usize) {
+        self.incoming_allow_list
+            .entry(peer)
+            .and_modify(|entry| *entry = entry.saturating_add(attempts))
+            .or_insert(attempts);
+    }
+
+    /// Remove some (or all) attempts of incoming connections from specified peer ID
+    pub(crate) fn remove_from_incoming_allow_list(
+        &mut self,
+        peer: &PeerId,
+        remove_attempts: Option<usize>,
+    ) {
+        if let Some(remove_attempts) = remove_attempts {
+            if let Some(attempts) = self.incoming_allow_list.get_mut(peer) {
+                *attempts = attempts.saturating_sub(remove_attempts);
+
+                if *attempts == 0 {
+                    self.incoming_allow_list.remove(peer);
+                }
+            }
+        } else {
+            self.incoming_allow_list.remove(peer);
+        }
+    }
+
+    /// Add to allow list some attempts of outgoing connections from specified peer ID that will bypass global limits
+    // TODO: Not using for now, but will be helpful upstream
+    #[allow(dead_code)]
+    pub(crate) fn add_to_outgoing_allow_list(&mut self, peer: PeerId, attempts: usize) {
+        self.outgoing_allow_list
+            .entry(peer)
+            .and_modify(|entry| *entry = entry.saturating_add(attempts))
+            .or_insert(attempts);
+    }
+
+    /// Remove some (or all) attempts of outgoing connections from specified peer ID
+    // TODO: Not using for now, but will be helpful upstream
+    #[allow(dead_code)]
+    pub(crate) fn remove_from_outgoing_allow_list(
+        &mut self,
+        peer: &PeerId,
+        remove_attempts: Option<usize>,
+    ) {
+        if let Some(remove_attempts) = remove_attempts {
+            if let Some(attempts) = self.outgoing_allow_list.get_mut(peer) {
+                *attempts = attempts.saturating_sub(remove_attempts);
+
+                if *attempts == 0 {
+                    self.outgoing_allow_list.remove(peer);
+                }
+            }
+        } else {
+            self.outgoing_allow_list.remove(peer);
+        }
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ConnectionHandler = <ConnectionLimitsBehaviour as NetworkBehaviour>::ConnectionHandler;
+    type ToSwarm = <ConnectionLimitsBehaviour as NetworkBehaviour>::ToSwarm;
+
+    fn handle_pending_inbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<(), ConnectionDenied> {
+        if let Some(peer) = remote_addr.iter().find_map(|protocol| {
+            if let Protocol::P2p(peer) = protocol {
+                Some(peer)
```

Comment:
This condition shouldn't work - we don't have a peer until we have a connection.